source/common/upstream/load_balancer_impl.cc

```diff
@@ -74,29 +93,46 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
   // 3 host sets with 20% / 20% / 10% health they will get 40% / 40% / 20% load to ensure total load
   // adds up to 100.
   const uint32_t total_health = std::min<uint32_t>(
-      std::accumulate(per_priority_health_.begin(), per_priority_health_.end(), 0), 100);
+      std::accumulate(per_priority_health.begin(), per_priority_health.end(), 0), 100);
   if (total_health == 0) {
     // Everything is terrible. Send all load to P=0.
     // In this one case sumEntries(per_priority_load_) != 100 since we sinkhole all traffic in P=0.
-    per_priority_load_[0] = 100;
+    per_priority_load[0] = 100;
     return;
   }
   size_t total_load = 100;
-  for (size_t i = 0; i < per_priority_health_.size(); ++i) {
+  for (size_t i = 0; i < per_priority_health.size(); ++i) {
     // Now assign as much load as possible to the high priority levels and cease assigning load when
     // total_load runs out.
-    per_priority_load_[i] =
-        std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
-    total_load -= per_priority_load_[i];
+    per_priority_load[i] =
+        std::min<uint32_t>(total_load, per_priority_health[i] * 100 / total_health);
+    total_load -= per_priority_load[i];
   }
   if (total_load != 0) {
     // Account for rounding errors.
-    ASSERT(total_load < per_priority_load_.size());
-    per_priority_load_[0] += total_load;
+    ASSERT(total_load < per_priority_load.size());
+    per_priority_load[0] += total_load;
   }
 }
 
-HostSet& LoadBalancerBase::chooseHostSet() {
+HostSet& LoadBalancerBase::chooseHostSet(LoadBalancerContext* context) {
+  if (context) {
+    const auto& pre_priority_filter = context->prePrioritySelectionFilter();
+
+    if (pre_priority_filter) {
+      // recompute per_priority_state with filter applied
```

Comment:
Why recompute vs. use this filter during initial `per_priority_state` construction?

Reply:
iirc the initial `per_priority_state` construction happens when the `LoadBalancer` is created, at which point we won't have access to the `LoadBalancerContext` that only comes into play when individual requests are being routed through the LB.